src/IterativeTrinary.php

```diff
@@ -0,0 +1,43 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan;
+
+use Iterator;
+
+final class IterativeTrinary
+{
+
+	/**
+	 * @param Iterator<TrinaryLogic> $it
+	 */
+	public static function and(Iterator $it): TrinaryLogic
+	{
+		$trinaries = [];
+		foreach ($it as $trinary) {
+			if (!$trinary->yes()) {
```

Comment:
This is only safe if the value is _no_

```suggestion
			if ($trinary->no()) {
```

Reply:
Yeah, the and case needs a few tests, and I think can also benefit from more early return logic